### DISCUSSION TITLE
[bug]: Getting a command not found error

### DISCUSSION THREAD

**Discussion by larsvanherwijnen** [spatie/laravel-pdf#101] [Q&A]
Hi all,

I am currently running into an issue when not specifying the nodeBinary and Npm binary. Is their any way to around this because this will not work on my teammates PC's and in production.

```
 The command "PATH=$PATH:/usr/local/bin:/opt/homebrew/bin NODE_PATH=`npm root -g` node 
'/Users/myuser/Herd/Bazaar/vendor/spatie/browsershot/src/../bin/browser.cjs' '{"url":"file:\/\/\/var\/tmp\/1025531601- 
0127108001709996159\/index.html","action":"pdf","options":{"path":"invoice.pdf","args":[],"viewport": 
{"width":800,"height":600},"displayHeaderFooter":false,"printBackground":true}}'" failed. Exit Code: 127(Command not found) 
Working directory: /Users/myuser/Herd/Bazaar/public Output: ================ Error Output: ================ 
sh: npm: command not found sh: node: command not found
```

Thanks!

**Comment by karamqubsi** [comment]
Hi,

I come across this error today, only managed to get it to work by modifying paths to node and npm binaries ( for browsershot )

is it the same problem on a live servers? didn't try that yet, I suspected it is just some limitation from Herd.

**Reply by Uhasith** [reply]
> Hi,
> 
> I come across this error today, only managed to get it to work by modifying paths to node and npm binaries ( for browsershot )
> 
> is it the same problem on a live servers? didn't try that yet, I suspected it is just some limitation from Herd.

could you find any solution for this? I'm trying to download a PDF inside my Livewire Component.

```
 public function downloadPdf()
    {
        $products = $this->products;
        return Pdf::view('pdf.purchase-order', compact('products'))->save('invoice.pdf');
    }
```

**Reply by Uhasith** [reply]
Also I tried this way. But Not working. keep given me this error. 

```
The command "PATH=$PATH:/usr/local/bin:/opt/homebrew/bin NODE_PATH=`npm root -g` node '/Users/hasithudayanga/Sites/empire/vendor/spatie/browsershot/src/../bin/browser.cjs' '{"url":"file:\/\/\/var\/tmp\/1596440366-0422314001723702342\/index.html","action":"pdf","options":{"path":"downloads\/purchase-order.pdf","args":[],"viewport":{"width":800,"height":600},"displayHeaderFooter":false,"printBackground":true}}'" failed. Exit Code: 1(General error)
```

```
   public function downloadPdf()
    {
        return Pdf::view('pdf.purchase-order', ['products' => $this->products])
            ->withBrowsershot(function (Browsershot $shot) {
                $shot->setNodeBinary('/Users/username/Library/Application Support/Herd/config/nvm/versions/node/v18.20.2/bin/node')
                ->setNpmBinary('/usr/local/lib/node_modules/npm');
            })
            ->save('purchase-order.pdf');
    }
 ```

**Reply by karamqubsi** [reply]
this should work ( similar change worked for me),  just make sure you are using correct paths , 

I think you may need to escape the space for `Application\ Support` and not sure why you are using other installation path for npm, 

so try this, ( change your username ) 
```php
public function downloadPdf()
    {
        return Pdf::view('pdf.purchase-order', ['products' => $this->products])
            ->withBrowsershot(function (Browsershot $shot) {
                $shot->setNodeBinary('/Users/username/Library/Application\ Support/Herd/config/nvm/versions/node/v18.20.2/bin/node')
                ->setNpmBinary('/Users/username/Library/Application\ Support/Herd/config/nvm/versions/node/v18.20.2/bin/npm');
            })
            ->save('purchase-order.pdf');
    }

```

**Comment by Uhasith** [comment]
> Hi all,
> 
> I am currently running into an issue when not specifying the nodeBinary and Npm binary. Is their any way to around this because this will not work on my teammates PC's and in production.
> 
> ```
>  The command "PATH=$PATH:/usr/local/bin:/opt/homebrew/bin NODE_PATH=`npm root -g` node 
> '/Users/myuser/Herd/Bazaar/vendor/spatie/browsershot/src/../bin/browser.cjs' '{"url":"file:\/\/\/var\/tmp\/1025531601- 
> 0127108001709996159\/index.html","action":"pdf","options":{"path":"invoice.pdf","args":[],"viewport": 
> {"width":800,"height":600},"displayHeaderFooter":false,"printBackground":true}}'" failed. Exit Code: 127(Command not found) 
> Working directory: /Users/myuser/Herd/Bazaar/public Output: ================ Error Output: ================ 
> sh: npm: command not found sh: node: command not found
> ```
> 
> Thanks!

Did you found any solution for this?

**Reply by juanborras** [reply]
Same problem here in production environment. Node & npm not found eventhough the work on the production command line.

Diggin deeper, the Browsershot class has the path for what it seems Mac and Un*x based systems...

protected string $includePath = '$PATH:/usr/local/bin:/opt/homebrew/bin';

Is there any priority when resolving the node binary on execution?

**Comment by CodeTechNL** [comment]
I ran into the same issue on generating a invoice. What helped for me is the following:

```

        $pdf = Pdf::view('notifications.financial.invoice_pdf')
            ->withBrowsershot(function (Browsershot $browsershot) {
                return $browsershot
                    // Target the whole dir as both node and NPM are there located
                    ->setIncludePath('~/.nvm/versions/node/v20.11.1/bin')
                    
                    // Or both by hand if they are in different folders
                    ->setNodeBinary('~/.nvm/versions/node/v20.11.1/bin/node')
                    ->setNpmBinary('~/.nvm/versions/node/v20.11.1/bin/npm')
                    ;
            })
            ->save('invoice.pdf');
```

Probably you can configure it somewhere globally.

**Comment by Khal-ID** [comment]
Any word on this? I'm having the same issue. I don't really wanna install homebrew.

**Reply by CodeTechNL** [reply]
You dont need homebrew, just set the correct paths for your Node and NPM

**Reply by Khal-ID** [reply]
I'm using Herd and nvm to manage node and npm versions

**Reply by CodeTechNL** [reply]
> I'm using Herd and nvm to manage node and npm versions

Are you on a Mac or Windows?

**Reply by Khal-ID** [reply]
> > I'm using Herd and nvm to manage node and npm versions
> 
> Are you on a Mac or Windows?

Mac

**Reply by CodeTechNL** [reply]
> > > I'm using Herd and nvm to manage node and npm versions
> > 
> > 
> > Are you on a Mac or Windows?
> 
> Mac

What if you run in the terminal the following?

> which node
/Users/******/.nvm/versions/node/v20.11.1/bin/node
> which npm
/Users/*******/.nvm/versions/node/v20.11.1/bin/npm

And then set the paths to the example I gave earlier:

```
 $pdf = Pdf::view('notifications.financial.invoice_pdf')
            ->withBrowsershot(function (Browsershot $browsershot) {
                return $browsershot
                    // Target the whole dir as both node and NPM are there located
                    ->setIncludePath('~/.nvm/versions/node/v20.11.1/bin')
                    
                    // Or both by hand if they are in different folders
                    ->setNodeBinary('~/.nvm/versions/node/v20.11.1/bin/node')
                    ->setNpmBinary('~/.nvm/versions/node/v20.11.1/bin/npm')
                    ;
            })
            ->save('invoice.pdf');
```

**Reply by Khal-ID** [reply]
```
> which node
/Users/khalid/Library/Application Support/Herd/config/nvm/versions/node/v22.14.0/bin/node
> which npm
/Users/khalid/Library/Application Support/Herd/config/nvm/versions/node/v22.14.0/bin/npm
```

It's worth noting that I'm not using browsershot directly, I'm using spatie/laravel-pdf which uses browsershot under the hood.

**Reply by CodeTechNL** [reply]
Yeah and the error is telling clearly: `sh: npm: command not found sh: node: command not found`. In other words: Browsershot is trying to locate the npm/node executable and can not find it, so you need to tell Browsershot where he can find your NPM and Node

**Reply by Khal-ID** [reply]
The issue is that browsershot isn't reading my path and choosing to go with the default:

`protected string $includePath = '$PATH:/usr/local/bin:/opt/homebrew/bin';`
